Implement pack_particles functions

gassmoeller opened the pull request:
This PR implements optimized pack functions for particles. As discussed in #8513 this is 20 times faster than the uncompressed general Utilities::pack function. Since the communication time was never a meaningful portion of the total runtime with this hand written routine, I would like to go back to it, instead of using the general one (which shows up as a significant portion of the runtime for my models).

marcfehling commented:
We should think about introducing optimized `pack`/`unpack` functions for `p::d::SolutionTransfer` (and probably `p::d::CellDataTransfer`) as well. I'll put that on my agenda.

gassmoeller commented:
Is this otherwise ready? I added some missing documentation.

masterleinad commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -38,6 +38,36 @@ DEAL_II_NAMESPACE_OPEN
+    /**
+     * This function unpacks a vector of bytes into a vector
+     * of particles. It assumes the vector of bytes was created
+     * with the pack_particles function above. If the particles
```
```suggestion
     * with the pack_particles() function above. If the particles
```

gassmoeller commented:
I modified the documentation following the suggestion.

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -28,6 +28,65 @@ DEAL_II_NAMESPACE_OPEN
+
+      while (data < &(*data_range.end()))
+        {
+          particles.push_back(Particle<dim, spacedim>(data, &property_pool));
```
```suggestion
          particles.emplace_back(data, &property_pool);
```

bangerth commented on source/particles/particle_handler.inst.in:
```diff
@@ -19,6 +19,21 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension : SPACE_DIMENSIONS)
+      namespace internal
+      \{
+        template std::vector<char>
+        pack_particles(
```
Do you actually need this? These are functions internal to the .cc file. We shouldn't need explicit instantiations. In fact, I think that these functions could also just be in an anonymous namespace.

gassmoeller commented:
You are right, currently the functions are not used outside of particle_handler.cc. I moved them into an anonymous namespace and removed the instantiation.

bangerth commented:
/rebuild
